// hanzo.js/lib/blueprints/deploy.js
// Generated by CoffeeScript 1.10.0
var byId, endpoint, fs, ref, statusCreated, statusNoContent, statusOk, upload;

fs = require('fs');

ref = require('../utils'), statusCreated = ref.statusCreated, statusNoContent = ref.statusNoContent, statusOk = ref.statusOk;

endpoint = function(x) {
  return "/site/" + x.siteId + "/deploy";
};

byId = function(x) {
  var ref1;
  return (endpoint(x)) + "/" + ((ref1 = x.id) != null ? ref1 : x.deployId);
};

upload = function(x) {
  return (byId(x)) + "/files/" + x.path;
};

module.exports = function(blueprints) {
  return blueprints.deploy = {
    create: {
      url: endpoint,
      method: 'POST',
      expects: statusCreated
    },
    update: {
      url: byId,
      method: 'PATCH',
      expects: statusOk
    },
    "delete": {
      url: byId,
      method: 'DELETE',
      expects: statusNoContent
    },
    restore: {
      url: byId,
      method: 'POST',
      expects: statusOk
    },
    upload: {
      url: upload,
      method: 'PUT',
      expects: statusOk,
      headers: {
        'Content-Type': 'application/octet-stream'
      },
      file: function(x) {
        return x.absolutePath;
      },
      followRedirects: true
    }
  };
};
